### PR TITLE
feat(component): add coalesce and generateFrames operators

### LINKED ISSUE #2446: @ngrx/component implement generateFrames observable projection function
# Problem 

**See demo [here](https://stackblitz.com/edit/ng-zone-af-tests)**

For the coalesce operator we use the animationFrame as durationSelector. 
RxJS in v7 provides an observable projection function called `animationFrames` that would solve the problem. However, Angular uses zone.js and zone.js patchen all async functions in the browser.

We needed to create an observable same as `animationFrames` form the unpatched `requestAnimationFrame` in an Angular project.

# Solution
This is the reason we need to create an observable projection function generic enough to take all kind of async producers from the browser.

**Signature:** 
```typescript
export function generateFrames(
  asyncProducer: asyncProducerFn,
  asyncCanceler: asyncCancelerFn,
  timestampProvider: TimestampProvider = Date
): Observable<number>;
```
**Usage:** 

```typescript
import { generateFrames } from '@ngrx/component';

const afUnpatched$ =  generateFrames(
  window.__zone_symbol__requestAnimationFrame,
  window.__zone_symbol__cancelAnimationFrame
);
```
**Todos:**
- [x] Implement default implementation based on RxJS animationFrames projection function
- [x] Implement configurable interval source
- [x] Implement tests for aditional logic
- [x] Create docs
  - [x] Ensure linking in TOCs

Related: #2441

## If accepted, I would be willing to submit a PR for this feature

[x] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

### PR DESCRIPTION
**Pull request by BioPhoton**
implement the operators, added tests, added docs, added diagrams, added rxjs-marbles to dev dependencies.

**Demonstrations for the need of these operators:**
- ~[generateFrames](https://stackblitz.com/edit/ng-zone-af-tests)~
- [coalesce](https://stackblitz.com/edit/ng-af-coalescing-tests)

closes ~#2446~ #2448 related to #2441

### CODE REVIEW

**Comment by BioPhoton** [comment]
@brandonroberts @alex-okrushko 

The tests are usingTestsScheduler directly now.

Let me know what to do to get this merged.

**Comment by BioPhoton** [comment]
@brandonroberts 

All todos from the updated section resolved.
This PR is ready to merge/review.

**Demonstrations for the need of these operators:**
- [generateFrames](https://stackblitz.com/edit/ng-zone-af-tests)
- [coalesce](https://stackblitz.com/edit/ng-af-coalescing-tests)

File: modules/component/spec/helper/observableMatcher.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import _ = require('lodash');
```
Comment (alex-okrushko) [inline modules/component/spec/helper/observableMatcher.ts:1]: ?

File: modules/component/src/core/operators/coalesce.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+const coalescingContextPropertiesMap = createPropertiesWeakMap<
+  object,
+  CoalescingContextProps
+>(ctx => ({
+  isCoalescing: false,
+}));
+
+const defaultCoalesceConfig: Pick<CoalesceConfig, 'leading' | 'trailing'> & {
+  context: undefined;
+} = {
```
Comment (alex-okrushko) [inline modules/component/src/core/operators/coalesce.ts:37]: this whole thing is just `CoalesceConfig`.

File: modules/component/src/core/operators/coalesce.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+): Pick<CoalesceConfig, 'leading' | 'trailing'> & {
+  context: object | undefined;
+} {
+  return {
+    ...defaultCoalesceConfig,
+    ...config,
+  };
+}
+
+export const defaultCoalesceDurationSelector = <T>(value: T) =>
```
Comment (alex-okrushko) [inline modules/component/src/core/operators/coalesce.ts:54]: value is not used?

File: modules/component/spec/core/utils/create_properties-weakmap.spec.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  getGlobalThis,
+} from '../../../src/core/utils';
+import { Injector } from '@angular/core';
+
+const propertyNameString = 'flag';
+const propertyNameSymbol = Symbol('flag');
+interface ManagedProperties {
+  flag: boolean;
+}
+let contextObject: any & object;
```
Comment (alex-okrushko) [inline modules/component/spec/core/utils/create_properties-weakmap.spec.ts:13]: that's just `any`

File: modules/component/src/core/operators/coalesce.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  generateFrames();
+
+/**
+ * @description
```
Comment (alex-okrushko) [inline modules/component/src/core/operators/coalesce.ts:58]: Please re-read your comments to correct some of the grammar.

File: modules/component/src/core/operators/coalesce.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+ * Limits the number of synchronous emitted a value from the source Observable to
+ * one emitted value per [`AnimationFrame`](https://developer.mozilla.org/en-US/search?q=AnimationFrame),
```
Comment (alex-okrushko) [inline modules/component/src/core/operators/coalesce.ts:60]: https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame ?
the search gives 10 results, none of which is particular interesting, except this one.

File: modules/component/spec/helper/observableMatcher.ts (hunk 1 above)
Comment (BioPhoton) [reply modules/component/spec/helper/observableMatcher.ts:1]: **blush**

File: modules/component/src/core/operators/coalesce.ts (hunk 2 above)
Comment (BioPhoton) [reply modules/component/src/core/operators/coalesce.ts:37]: Yes!

File: modules/component/src/core/operators/coalesce.ts (hunk 6 above)
Comment (BioPhoton) [reply modules/component/src/core/operators/coalesce.ts:60]: Good catch!

File: modules/component/src/core/operators/coalesce.ts (hunk 3 above)
Comment (BioPhoton) [reply modules/component/src/core/operators/coalesce.ts]: deleted with generate frames

File: modules/component/spec/core/utils/create_properties-weakmap.spec.ts (hunk 4 above)
Comment (BioPhoton) [reply modules/component/spec/core/utils/create_properties-weakmap.spec.ts]: fixed!

**Comment by BioPhoton** [comment]
Big thanks to @hoebbelsB for spotting a nasty bug!

File: modules/component/src/core/operators/coalesce.ts (hunk 5 above)
Comment (BioPhoton) [reply modules/component/src/core/operators/coalesce.ts:50]: Done and rephreased.

**Comment by BioPhoton** [comment]
@alex-okrushko it helps to diff the code of coalesce.ts with [throttle](https://github.com/ReactiveX/rxjs/blob/master/src/internal/operators/throttle.ts)

I just added a view lines. Also, the new tests should help.

**Comment by brandonroberts** [comment]
@BioPhoton if this is answered somewhere else, point me to it, but from looking, I don't see any Angular-specific code here, so why couldn't this be pushed upstream to RxJS?

**Comment by BioPhoton** [comment]
Because we need coalesceing in our pipe/directive.

This is essential for the performance and we discussed it multiple times.

I created a click demo and a design doc. Tested it and informed the team regunarly and miltiple times in the chat and other PRs.

Honestly I'm confused about the question.

**Comment by BioPhoton** [comment]
@sanderelias

**Comment by BioPhoton** [comment]
Also, as mentioned in slack, my big achievements of kick out zone completely from the code base was the reason to also kick out the generateFrames operator.

**Comment by BioPhoton** [comment]
All those i for are collected in the design docs I was asked to create several month ago.

**Comment by BioPhoton** [comment]
https://hackmd.io/42oAMrzYReizA65AwQ7ZlQ

**Comment by BioPhoton** [comment]
I remenber I spent time to document the related strategy.
Here the comments:

# Local Strategy

This strategy is rendering the actual component and
all it's children that are on a paththat is marked as dirty or has components with `ChangeDetectionStrategy.Default`.

As detectChanges has no coalescing of render calls
like `ChangeDetectorRef#markForCheck` or `ɵmarkDirty` has, so we have to apply our own coalescing, 'scoped' on component level.

Coalescing, in this very manner,
means **collecting all events** in the same [EventLoop](https://developer.mozilla.org/de/docs/Web/JavaScript/EventLoop) tick,
that would cause a re-render and execute **re-rendering only once**.
 
'Scoped' coalescing, in addition, means **grouping the collected events by** a specific context.
E. g. the **component** from which the re-rendering was initiated.

 | Name        | ZoneLess VE/I | Render Method VE/I  | Coalescing VE/I  |
 |-------------| --------------| ------------------------- | ------------------ |
 | `local`        | ✔️/✔️          | dC / ɵDC                      | ✔️ + C/ LV          |

@param config { StrategyFactoryConfig } - The values this strategy needs to get calculated.
@return {CdStrategy<T>} - The calculated strategy

```typescript
export function createLocalStrategy<T>(
  config: StrategyFactoryConfig
): CdStrategy<T> {
  const durationSelector = getZoneUnPatchedDurationSelector();

  const coalesceConfig: CoalesceConfig = {
    // @TODO ensure that context is === to _lView across class and template (all cases!!!)
    // If yes, kick out _lView
    context: (IS_VIEW_ENGINE_IVY
      ? (config.cdRef as any)._lView
      : (config.cdRef as any).context) as any,
  };

  function render() {
    // @TODO ensure that detectChanges is behaves identical to ɵdetectChanges
    // If yes, kick out ɵdetectChanges
    if (IS_VIEW_ENGINE_IVY) {
      detectChanges((config.cdRef as any).context);
    } else config.cdRef.detectChanges();
  }

  const behaviour = () => (o$: Observable<T>): Observable<T> => {
    return o$
      .pipe(coalesce(durationSelector, coalesceConfig));
  };

  return {
    behaviour,
    render,
    name: 'local',
  };
}
```

---

I kept the comments so you see the missing tests I have to do to make it independent of internal Angular APIs.

**Comment by BioPhoton** [comment]
Please let me know about further needed information. I'm happy to create another document if needed. 

Just let me know how to proceed.

**Comment by BioPhoton** [comment]
> so why couldn't this be pushed upstream to RxJS

I can see if this is possible. Closing this one.

**Pull request closed without merging**